### ISSUE TITLE
Couldn't demux in Linux seedbox without root

### ISSUE THREAD

**Issue by a-nunes** [TMD20/Remuxer#1]
I couldn't demux movies on my seedbox, it requires wine to run eac3to, but I only can access it through SSH and wine seems to try to open a window display.
I've been looking for a solution to this problem, but not with eac3to, instead I've found DGDemux for linux, and it works perfectly.
Would it be possible to turn PTMuxer a seedbox friendly by replacing eac3to with wine for dgdemux for linux?

You can find it [here](https://forum.doom9.org/showthread.php?t=182842).

Thank you very much.

**Comment by TMD20** [comment]
What is your seedbox provider

**Comment by a-nunes** [comment]
> What is your seedbox provider

Ultra.cc

**Comment by TMD20** [comment]
I was going to suggest docker, but it doesn't look like they allow custom images.

I'm not sure of the acceptance of dgdemux on trackers. I know one tracker allows it for 2160p in some cases. I think split playlist

I have to see if it has a command line and if it does I should be able to add it at some point. It will probably just be a command line argument to force it usage. 

In the future it was going to be automatically used if one encountered a 2160p split playlist. I haven't had any disk with this feature so I've been unable to test it

**Comment by a-nunes** [comment]
Oh, I see... It's linux version works only on CLI, I got this from its help output:

---
DGDemux 1.0.0.58 (linux) by Donald A. Graft
Copyright (C) 2019-2021 Donald A. Graft, All Rights Reserved [Made in USA]
(dgdemux -h for help)

DGDemux is a standalone demuxer for BluRay/UHD disks, including 3D BluRays.
It takes a playlist and demuxes all the elementary streams (audio and video),
the subtitle streams, and the chapters. The demuxed files are all
in sync with the demuxed video, even for seamless branching titles.

Basic Usage
-----------

You must have a full decrypted disk structure on local storage, obtained
from (for example) AnyDVD HD or MakeMKV. Then the usage is:

DGDemux -i full_path_to_MPLS_file -o full_path_to_destination_with_filename_suffix

Example
-------

DGDemux -i H:/Rips/CARS_2/BDMV/PLAYLIST/00800.mpls -o d:/tmp/CARS_2/00800

Note that for the -i, -o, -d, and -l options, if there is any white space in the
path, the entire path string should be enclosed in quotes, e.g.:

DGDemux -i "H:/Rips/CARS 2/BDMV/PLAYLIST/00800.mpls"

Command Line Options
--------------------

-v: Print version information.

-h: Print help information.

-ang x: Selects angle given by integer x. The numbering starts at 1.

-fo: Demux forced subtitles only.

-d path: This option specifies the disk to be processed. It should be a
full path to the desired top-level directory of the decrypted disk structure.
DGDemux will output a list of playlists with their sizes. This must be the last
option given on the command line.

-i path: This option specifies the playlist to be processed. It should be a
full path to the desired playlist file in the decrypted disk structure.

-o path: This a full path to the directory and name to be used for the demuxed files.
Please do not include an extension. The name at the end of the path will be used to
name the demuxed files. This is typically the first part of the MPLS filename.
In the example above, the path "D:/tmp/CARS_2/00800" specifies the output
directory as "D:/tmp/CARS_2/", and specifies the prefix "00800" for the names of
the demuxed files. Be careful when using multiple instances of DGDemux that the
name prefix is different so that the instances cannot try to create files with the
same filename.

If the -o option is omitted, then a list of available streams to demux will be output.

-l path_to_file: Redirect output to the specified file. This is used by DGDemuxGUI.
You can also use it to discard the output when using DGDemux directly: give
'-l nul' on the command line.

-demux list: list is a comma-separated list of streams to be demuxed. For example,
"-demux 1011,1100". The numbers are the hex values of the corresponding PIDs, as shown
when a listing is generated by omitting the -o option. Also note that if this option is
omitted, then all the streams will be demuxed. If you want chapters to be demuxed, then
you must include "Chapters" in the PID list, for example, "-demux 1011,1100,Chapters"

-sf: Skip the first M2TS file in the playlist. Some disks, such as Naked Alibi, have a
very short first M2TS file in the playlist that is typically a logo without audio.
The lack of audio causes a parsing failure upon opening the MPLS, such that the
audio is not detected for the title. To process such disks, you can give the -sf
option. The chapters will be adjusted accordingly.

-sl: Skip the last M2TS file in the playlist.

-nsthd: The 'no split thd' option demuxes THD with it's embedded AC3 as a single THD+AC3
stream. When this option is omitted, THD is demuxed into two separate streams, one THD
without embedded AC3, and one AC3. Note that file gaps processing for the THD and embedded
AC3 streams is performed only when this option is not given, i.e., when the streams are
split.

-pre: If this option is given DGDemux tries to preallocate disk space for demuxing
in order to reduce fragmentation of the demuxed files. In low disk space situations, this
may cause demux failure even though there is (barely) enough space. For this situation,
do not give this option.

-sp: Strict playlist. This option removes garbage at the start and end of a playlist.
Garbage refers to content in a playlist that is outside the range defined
by the IN_time of the first M2TS and the OUT_time of the last M2TS. This is distinct from
-sf and -sl. The IN_time of an M2TS need not correspond to the first video access unit of
an M2TS. For example, CROUPIER (one M2TS in the playlist) has almost a full minute of color
bars and blackness before the IN_time, and then some more blackness after the OUT_time.
Note: This feature is currently experimental. Please report your results with it.

-ch: Chop excess audio. Within each M2TS file, remove audio that specifies a PTS after the
OUT_time of the M2TS (excess audio).

-nogaps: If this option is checked file gaps processing for the audio is disabled. Typically used
for research and debugging.

-ep: Episode demuxing mode. Each M2TS in the playlist is demuxed into separate files. When in this
mode, the progress indication is of the form "[n] x", where n is the episode number (1-based)
and x is the progress percentage for that episode.

-sk n: Skip the first n M2TS files. Do not use with any other skip options. This option is
intended for use transparently by the -ep option.

Copyright (C) 2019-2021 Donald A Graft, All Rights Reserved

**Comment by TMD20** [comment]
This is close to being implemented

I had to rewrite a bunch of code, as it would not handle dgdemux prior.

once the dgdemux branch reaches feature parity with the main branch, then I will update the main branch with the modified code.

It is mainly just splitplaylist mode that need to be supported.

**Comment by a-nunes** [comment]
Hey, many thanks to all your work. It is awesome that someone is working on it! =D

**Comment by TMD20** [comment]
I updated the main branch
with dgdemux support

**Comment by a-nunes** [comment]
Hey, what's up?

I've checked it with Movies and it worked for me. I didn't check it for TV Shows yet.
Here are a few minor changes that I've done in order to make it work properly in my box:

1. Disabled tesseract (optional)
2. Got this error on my first run:
`OSError: ./PTMuxer/mediatools/BDSupReader/src/RunLength.so: cannot open shared object file: No such file or directory`
Solved it by removing the folder and cloning the original repo here: [BDSupReader](https://github.com/Sec-ant/BDSupReader)
3. This was the next error:
`ModuleNotFoundError: No module named 'bufferedrandomplus'`
Looks like the local imports didn't work correctly, so I just add a `.` before the imports on this file `PTMuxer/mediatools/BDSupReader/src/bdsupreader.py`.
Let them like this:
 ```
import os, logging
import numpy as np
from .bufferedrandomplus import BufferedRandomPlus
from .imageutil import *
from .timeutil import *
from .enumerations import *
```
4. Next error was:
`[Errno 2] No such file or directory: './PTMuxer/binaries/wine/wine'`
As said on my opening issue, wine doesn't seems to work on a headless environment. Solved by doing:
```
mkdir -p PTMuxer/binaries/wine
ln -s /usr/bin/mono PTMuxer/binaries/wine/wine
```
5. The next one:
`[Errno 18] Invalid cross-device link: '/tmp/PTMuxer_a83gu8ch/. PID 1011.264' -> '001.h264'`
After some investigation I've found out that I don't have permission to write on `/tmp`, so I just hardcoded a tmp folder on `config.py` file:
from: `tempFolder=tempfile.gettempdir()`
to: `tempFolder='/home/my_user/tmp'`

6. The last one:
`Cannot open assembly './PTMuxer/binaries/bdsup2sub.exe': File does not contain a valid CIL image.`
This one happens because the 'wine' used actually is a softlink to 'mono', and bdsup2sub.exe isn't a C# binary. Turns out that this step looks like related to OCR and doesn't affect directly the demux / remux process.

After doing this changes, I was able to make the whole process of demuxing and remuxing a movie on my box without root. 
Thank you very much for all your effort in making this possible.

**Comment by TMD20** [comment]
Thank you for you report

You seem to have some technical proficiency, but I definitively want to make this as easy as possible for less experience users.
So your feedback is very helpful.
`
The part about the temp folders was already included in the documentation in the advanced section. It seems like maybe you read that before changing that setting. Also it is in a easy to find place.`

`I ran into the same error with bdsupreader, but it seems you found a better way to bypass it so I will see about implemented that.
`

Also it's great that you were able to get this to work I was worried some of the binaries may cause issues.
```
bdsup2sub is actually for forced that are embedded in a track, it will separate the forced subtitles on that track 
However from what I've heard it is usually a separate track. So not strictly required, and the demux process can complete with out it. 
forcedsubs: https://docs.google.com/spreadsheets/d/1nAMVVnmk3vSaUiuFXcN7D0R69FHIP4tfIkTdThTkXEk/edit#gid=2134580265

There are actually two version of the app
the original java version
A exe rewrite 

I set linux system to use the java version by default.
Java install is 100mb so it not something I want to include in the repo
But it is fairly easy to install/remove
```

**Comment by a-nunes** [comment]
I saw that you've made some updates on main branch. I'll test it again ASAP and let you a new feedback, this way we can help others to make it work with less effort.

**Comment by a-nunes** [comment]
> I saw that you've made some updates on main branch. I'll test it again ASAP and let you a new feedback, this way we can help others to make it work with less effort.

I've made a pull and merged the updates with my local ones. It worked perfectly.

**Comment by TMD20** [comment]
Thanks for your support
I've added in all the changes that you made in your local version to the repo.

**Mono**
I spent some time looking for mono, but I have not been able to find anything prebuilt. I'm not sure if your seedbox, just had this preinstalled.
To me it looks like it might because the path your gave was not in the home directory.

**Sudoless Wine and Mono**
I had been looking for a way to wine or mono without sudo, but it seemed like a complete and ready made solution did not exists, but I was able to finally find one yesterday.
https://github.com/Kron4ek/Conty

This should of course work with any program that needs wine on Linux. This includes eac3to and bdinfo from my testing. Additionally since it just a self contained script there is no need for sudo access to run. 

The download is 1.7GB which is quite a lot, so this will be completely optional like java
